src/rust/engine/process_execution/src/lib.rs

```diff
@@ -1093,35 +1143,48 @@ pub async fn make_execute_request(
     append_only_caches_base_path: Option<&str>,
 ) -> Result<EntireExecuteRequest, String> {
     const WRAPPER_SCRIPT: &str = "./__pants_wrapper__";
+    const SANDBOX_ROOT_TOKEN: &str = "__PANTS_SANDBOX_ROOT__";
 
     // Implement append-only caches by running a wrapper script before the actual program
     // to be invoked in the remote environment.
-    let wrapper_script_digest_opt = match (append_only_caches_base_path, &req.append_only_caches) {
-        (Some(base_path), caches) if !caches.is_empty() => {
-            let script = make_wrapper_for_append_only_caches(
-                caches,
-                base_path,
-                req.working_directory.as_ref().and_then(|p| p.to_str()),
-            )?;
-            let digest = store
-                .store_file_bytes(Bytes::from(script), false)
-                .await
-                .map_err(|err| {
-                    format!("Failed to store wrapper script for remote execution: {err}")
-                })?;
-            let path = RelativePath::new(Path::new(WRAPPER_SCRIPT))?;
-            let snapshot = store.snapshot_of_one_file(path, digest, true).await?;
-            let directory_digest = DirectoryDigest::new(snapshot.digest, snapshot.tree);
-            Some(directory_digest)
-        }
-        _ => None,
+    let wrapper_script_content_opt = {
+        let sandbox_root_token = req
+            .argv
+            .iter()
+            .any(|arg| arg.contains("{chroot}"))
+            .then_some(SANDBOX_ROOT_TOKEN);
+        maybe_make_wrapper_script(
+            &req.append_only_caches,
+            append_only_caches_base_path,
+            req.working_directory.as_ref().and_then(|p| p.to_str()),
+            sandbox_root_token,
+        )?
+    };
+    let wrapper_script_digest_opt = if let Some(script) = wrapper_script_content_opt {
+        let digest = store
+            .store_file_bytes(Bytes::from(script), false)
+            .await
+            .map_err(|err| format!("Failed to store wrapper script for remote execution: {err}"))?;
+
+        let path = RelativePath::new(Path::new(WRAPPER_SCRIPT))?;
+        let snapshot = store.snapshot_of_one_file(path, digest, true).await?;
+        let directory_digest = DirectoryDigest::new(snapshot.digest, snapshot.tree);
+        Some(directory_digest)
+    } else {
+        None
     };
 
     let arguments = match &wrapper_script_digest_opt {
         Some(_) => {
             let mut args = Vec::with_capacity(req.argv.len() + 1);
             args.push(WRAPPER_SCRIPT.to_string());
-            args.extend(req.argv.iter().cloned());
+            for orig_arg in &req.argv {
+                if orig_arg.contains("{chroot}") {
+                    args.push(orig_arg.replace("{chroot}", SANDBOX_ROOT_TOKEN));
+                } else {
+                    args.push(orig_arg.clone());
+                }
```

Comment:
FWIW, I suspect this code could unconditionally replace, which should be pretty similar to doing the explicit search that finds nothing (when there's no results), I imagine.

```suggestion
                args.push(orig_arg.replace("{chroot}", SANDBOX_ROOT_TOKEN))
```

One could potentially code-golf down to something similar to the original code: swap the `cloned()` for a `map(...)`:

```rust
let args = ...;
args.push(...);
args.extend(req.argv.iter().map(|orig_arg| orig_arg.replace("{chroot}", SANDBOX_ROOT_TOKEN)))
```

Reply:
Nice. Done.